pslab/external/ccs811.py

```diff
@@ -0,0 +1,144 @@
+    _REG_STATUS_BIT_FW_MODE = 7
+    _REG_STATUS_BIT_APP_ERASE = 6
+    _REG_STATUS_BIT_APP_VERIFY = 5
+    _REG_STATUS_BIT_APP_VALID = 4
+    _REG_STATUS_BIT_DATA_READY = 3
+    # 2:1 -
+    _REG_STATUS_BIT_ERROR = 0
+
+    def __init__(self, **args):
+        self._ADDRESS = args.get('address', self._ADDRESS)
+        super().__init__(self._ADDRESS)
+        self.fetchID()
+        self.softwareReset()
+
+    def softwareReset(self):
+        self.write([0x11, 0xe5, 0x72, 0x8a], self._SW_RESET)
+
+    def fetchID(self):
+        hardware_id = (self.read(1, self._HW_ID))[0]
+        print(f'hex(hardware_id) = {hex(hardware_id)}')
+        time.sleep(0.02)  # 20ms
+        hardware_version = (self.read(1, self._HW_Version))[0]
+        print(f'hex(hardware_version) = {hex(hardware_version)}')
+        time.sleep(0.02)  # 20ms
+        boot_version = (self.read(2, self._FW_Boot_Version))[0]
+        print(f'hex(boot_version) = {hex(boot_version)}')
+        time.sleep(0.02)  # 20ms
+        app_version = (self.read(2, self._FW_App_Version))[0]
+        print(f'hex(app_version) = {hex(app_version)}')
+
+    def appErase(self):
+        ignore = self.write([0xE7, 0xA7, 0xE6, 0x09], self._APP_ERASE)
+        time.sleep(0.3)
+
+    def appStart(self):
+        ignore = self.write([], self._APP_START)
+
+    def setMeasureMode(self, mode):
+        self.write([mode << 4], self._MEAS_MODE)
+
+    def getMeasureMode(self):
+        print(self.read(10, self._MEAS_MODE))
+
+    def getStatus(self):
+        status = (self.read(1, self._STATUS))[0]
+        return status
+
+    def decodeStatus(self, status):
+        s = ''
+        if (status & (1 << self._REG_STATUS_BIT_FW_MODE)) > 0:
+            s += "Sensor is in application mode"
+        else:
+            s += "Sensor is in boot mode"
+        if (status & (1 << self._REG_STATUS_BIT_APP_ERASE)) > 0:
+            s += ", APP_ERASE"
+        if (status & (1 << self._REG_STATUS_BIT_APP_VERIFY)) > 0:
+            s += ", APP_VERIFY"
+        if (status & (1 << self._REG_STATUS_BIT_APP_VALID)) > 0:
+            s += ", APP_VALID"
+        if (status & (1 << self._REG_STATUS_BIT_DATA_READY)) > 0:
+            s += ", DATA_READY"
+        if (status & (1 << self._REG_STATUS_BIT_ERROR)) > 0:
+            s += ", ERROR"
+        return s
+
+    def measure(self):
+        data = self.read(8, self._ALG_RESULT_DATA)
+        eCO2 = data[0] * 256 + data[1]
+        eTVOC = data[2] * 256 + data[3]
+        status = data[4]
+        error_id = data[5]
+        raw_data = 256 * data[6] + data[7]
+        raw_current = raw_data >> 10
+        raw_voltage = (raw_data & ((1 << 10) - 1)) * (1.65 / 1023)
+
+        result = {"eCO2": eCO2, 'eTVOC': eTVOC, 'status': status, 'error_id': error_id}
+
+        if error_id == 0:
+            pass
+        elif error_id == 1:
+            print("The CCS811 received an I²C read request to a mailbox ID that is invalid")
+        else:
+            print("Other error")
```

Comment:
These should `raise RuntimeError`. The user can decide whether to catch the exception.

Reply:
Fixed in 6b00a77.